cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -403,6 +467,70 @@ public void run() {
     }
   }
 
+  /**
+   * Serializes a {@link CredentialAccessBoundary} object into Protobuf wire format.
+   */
+  private byte[] serializeCredentialAccessBoundary(
+      CredentialAccessBoundary credentialAccessBoundary) throws IOException {
+    List<AccessBoundaryRule> rules =
+        credentialAccessBoundary.getAccessBoundaryRules();
+    ClientSideAccessBoundary.Builder accessBoundaryBuilder =
+        ClientSideAccessBoundary.newBuilder();
+
+    for (AccessBoundaryRule rule : rules) {
+      ClientSideAccessBoundaryRule.Builder ruleBuilder =
+          accessBoundaryBuilder.addAccessBoundaryRulesBuilder()
+              .addAllAvailablePermissions(rule.getAvailablePermissions())
+              .setAvailableResource(rule.getAvailableResource());
+
+      if (rule.getAvailabilityCondition() != null) {
+        String availabilityCondition =
+            rule.getAvailabilityCondition().getExpression();
+
+        Expr availabilityConditionExpr = this.compileCel(availabilityCondition);
+        ruleBuilder.setCompiledAvailabilityCondition(availabilityConditionExpr);
+      }
+    }
+
+    return accessBoundaryBuilder.build().toByteArray();
+  }
+
+  /**
+   * Compiles CEL expression from String to an {@link Expr} proto object. 
+   */
+  private Expr compileCel(String expr) throws IOException {
+    try {
+      CelAbstractSyntaxTree ast = celCompiler.parse(expr).getAst();
+
+      CelProtoAbstractSyntaxTree astProto =
+          CelProtoAbstractSyntaxTree.fromCelAst(ast);
+
+      return astProto.getExpr();
+    } catch (CelValidationException exception) {
+      throw new IOException("Failed to parse CEL expression: " +
+                            exception.getMessage());
+    }
+  }
+
+  /**
+   * Encrypts the given bytes using a sessionKey using Tink Aead.
+   */
+  private byte[] encryptRestrictions(byte[] restriction, String sessionKey) throws InternalError {
+    try {
+      byte[] rawKey = Base64.getDecoder().decode(sessionKey);
+
+      KeysetHandle keysetHandle = TinkProtoKeysetFormat.parseKeyset(
+          rawKey, InsecureSecretKeyAccess.get());
+
+      Aead aead =
+          keysetHandle.getPrimitive(RegistryConfiguration.get(), Aead.class);
+
+      return aead.encrypt(restriction, /*associatedData=*/new byte[0]);
+    } catch (GeneralSecurityException exception) {
+      throw new InternalError("Failed to parse keyset: " + exception.getMessage());
```

Comment:
qq, what's the benefit of re-throwing this as InternalError instead of keeping it as GeneralSecurityException?

Reply:
Changed it to throw GeneralSecurityException

Reply:
On second thought, do you know what the error message is from the `GeneralSecurityException` when it fails? If it's too vague (i'm assuming it may be because it says General), perhaps we can re-throw with the error message you wrote above. If we do re-throw, I'm not too certain about InternalError. I think a different exception that would highlight the error type would be best.

Reply:
The message is "Parse keyset failed". I think it's specific.